crates/core/src/db/relational_db.rs

```diff
@@ -121,7 +198,21 @@ impl RelationalDB {
             move || durability.size_on_disk()
         });
 
-        Self::open(root, address, Some((durability.clone(), disk_size_fn)))?.apply(durability)
+        let snapshot_dir = root.as_ref().join("snapshots");
+        create_dir_all(&snapshot_dir)?;
+        let snapshot_repo = SnapshotRepository::open(snapshot_dir, address, instance_id).map(Arc::new)?;
+
+        let start = std::time::Instant::now();
+        let res = Self::open(
+            root,
+            address,
+            Some((durability.clone(), disk_size_fn)),
+            Some(snapshot_repo),
+        )?
+        .apply(durability);
```

Comment:
I think we will need to determine the `ConnectedClients` entirely from `st_clients`, as it appears right after this point.

Consider: the database crashed right after a snapshot, so the history suffix is empty -- but there might be connected clients in `st_clients`. Before we had snapshotting, an empty history meant an empty database.

Reply:
I believe this is the case now that I've "rebased" onto latest master, including `st_clients`. Please verify that I'm doing this correctly.

Reply:
`apply_history` returns the unmatched `__connect__`s during replay -- but what I meant was, what if the snapshot's tx offset is the last offset in the commitlog (hence the history is empty)?

Arguably an edge case, so fine by me to merge an fix in a later patch (I can submit one if you want).

Reply:
Oh, I assumed it would read the connected clients out of the table. I thought we had done away with groveling the logs for connected clients. I think a follow-up would be good, and if you're willing to do so, that would be great!

Reply:
https://github.com/clockworklabs/SpacetimeDB/pull/1366